### ISSUE TITLE
[visualforce] NPE when analyzing standard field references in visualforce page

### ISSUE THREAD

**Issue by Daniel-Ventura-25** [pmd/pmd#5476]
[visualforce] Parse Error: NullPointerException in Visualforce Rules

**Affects PMD Version:**
7.7.0

**Description:**
We encountered a parse error when running Visualforce rules on specific Visualforce pages. The error occurs when the page contains a standard field reference, such as {!Case.Description}. However, if we use a custom field, such as {!Case.Description__c}, no error is triggered.

The issue is intermittent and only reproducible when running PMD Visualforce rules on a large set of Visualforce pages (approximately 7,000). Running the rules on the specific problematic file alone does not trigger the error.

**Error Message:
The following error is encountered:**

```
NullPointerException: Cannot invoke "org.w3c.dom.Node.getNodeValue()" because "typeNode" is null
```
**Exception Stacktrace:**

```
<error filename=".\force-app\main\default\pages\nameofthepage.page" msg="NullPointerException: Cannot invoke &quot;org.w3c.dom.Node.getNodeValue()&quot; because &quot;typeNode&quot; is null">
<![CDATA[java.lang.NullPointerException: Cannot invoke "org.w3c.dom.Node.getNodeValue()" because "typeNode" is null
	at net.sourceforge.pmd.lang.visualforce.ast.ObjectFieldTypes.parseSfdxCustomField(ObjectFieldTypes.java:175)
	at net.sourceforge.pmd.lang.visualforce.ast.ObjectFieldTypes.findDataType(ObjectFieldTypes.java:129)
	at net.sourceforge.pmd.lang.visualforce.ast.SalesforceFieldTypes.getDataType(SalesforceFieldTypes.java:70)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:142)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:26)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTElExpression.acceptVfVisitor(ASTElExpression.java:14)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:21)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTAttributeValue.acceptVfVisitor(ASTAttributeValue.java:15)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:20)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTAttribute.acceptVfVisitor(ASTAttribute.java:53)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:19)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:89)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:26)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTElement.acceptVfVisitor(ASTElement.java:82)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:9)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTContent.acceptVfVisitor(ASTContent.java:15)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:19)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:89)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:26)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTElement.acceptVfVisitor(ASTElement.java:82)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:9)
	at net.sourceforge.pmd.lang.visualforce.ast.ASTContent.acceptVfVisitor(ASTContent.java:15)
	at net.sourceforge.pmd.lang.visualforce.ast.AbstractVfNode.acceptVisitor(AbstractVfNode.java:32)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitChildren(AstVisitorBase.java:31)
	at net.sourceforge.pmd.lang.ast.AstVisitorBase.visitNode(AstVisitorBase.java:38)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visitVfNode(VfVisitor.java:7)
	at net.sourceforge.pmd.lang.visualforce.ast.VfVisitor.visit(VfVisitor.java:8)
	at net.sourceforge.pmd.lang.visualforce.ast.VfExpressionTypeVisitor.visit(VfExpressionTypeVisitor.java:63)
	at net.sourceforge.pmd.lang.visualforce.ast.VfParser.parseImpl(VfParser.java:37)
	at net.sourceforge.pmd.lang.visualforce.ast.VfParser.parseImpl(VfParser.java:16)
	at net.sourceforge.pmd.lang.ast.impl.javacc.JjtreeParserAdapter.parse(JjtreeParserAdapter.java:36)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.parse(PmdRunnable.java:112)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:132)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
]]>
</error>
```
**Code Sample demonstrating the issue:**

```
<apex:page standardController="Case">
    <apex:outputField value="{!Case.Description}"/>
</apex:page>
```
The error occurs at line 2.
If {!Case.Description} is replaced with {!Case.Description__c}, the error does not occur.
Steps to reproduce:

1. Create a Visualforce page with the above code snippet containing a reference to a standard field, such as {!Case.Description}.
2. Run PMD Visualforce rules on a large set of Visualforce pages (e.g., ~7,000 pages).
3. Observe that a NullPointerException occurs during the PMD analysis.

Note: The issue does not occur if the PMD rules are executed only on the specific example provided above. The error is only reproducible in a large-scale analysis scenario.

**Additional Information:**
We observed four similar parse errors across different Visualforce pages, all related to standard field references like {!Case.Description}.

**Running PMD through:**
CLI

**JDK and OS Information:**

**JDK Version:** 
openjdk 21.0.4 2024-07-16 LTS
OpenJDK Runtime Environment Microsoft-9889606 (build 21.0.4+7-LTS)
OpenJDK 64-Bit Server VM Microsoft-9889606 (build 21.0.4+7-LTS, mixed mode, sharing)

**OS:** Microsoft Windows

**Comment by adangel** [comment]
PMD looks at a files `...field-meta.xml` to figure out, which custom fields exist.
Can you share your complete directory tree - including the xml files in addition to the "*.page" file?

From the exception stacktrace, it looks like these meta xml files exist, but do not have a correct content...

Note: This is all about sfdx projects, which have a very specific directory layout.
Note2: The feature that is having a problem is type resolution: https://docs.pmd-code.org/latest/pmd_languages_visualforce.html#type-resolution

**Comment by Daniel-Ventura-25** [comment]
Hi @adangel,

When we run only the failing pages, our directory structure looks like this:

## Directory Structure

```plaintext
testFolder
└── force-app
    └── main
        └── default
            └── pages
                ├── CasePrintPage.page
                ├── CasePrintPage.page-meta.xml
                ├── CasePrintPageTemplate.page
                ├── CasePrintPageTemplate.page-meta.xml
                ├── eLisaPage.page
                ├── eLisaPage.page-meta.xml
                ├── RD_CC_CasePrintPage.page
                ├── RD_CC_CasePrintPage.page-meta.xml
                ├── RD_CC_CasePrintPageTemplate.page
                ├── RD_CC_CasePrintPageTemplate.page-meta.xml

```

For the full package (where we encounter the error), the structure remains the standard SFDX format, retrieved as usual. It follows the same directory structure as above but includes all components. 
Additionally, the root folder contains the sfdx specific files, as sfdx-project.json file.

I just copied all the files from our full package structure (including sfdx-project.json, .sfdx folder, etc.) to this test folder, but we're still not seeing the same errors that occur when running PMD inside the full package.

Thank you so much for your help! I really appreciate it!  
Let me know if you need any additional details!

**Comment by Daniel-Ventura-25** [comment]
Hi @adangel !

I have some updates!

After further investigation, we identified the root cause of the issue with the four Visualforce pages. These pages reference Case.Description, and PMD is attempting to check for the "type" tag on this field.

However, since Description is a standard field, it does not include a "type" tag. As a result, when the code tries to access it using "ObjectFieldTypes.findDataType" (as seen in the stack trace), it fails because standard fields do not have a defined data type in this context, leading to a Null Pointer Exception.

Let me know your thoughts or if you have any suggestions on how we should proceed.

Best,
Daniel Ventura

**Comment by adangel** [comment]
Thanks for the investigations!

From what I understand: `Case` is a standard object and not a custom object. It is documented here: https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_case.htm
It has various standard fields, one of them is `Description`.

When you save your sfdx project as [Salesforce DX Project Structure and Source Format](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm) with decomposed metadata, then there is a file created `.../objects/Case/fields/Description.field-meta.xml` that has a different content than e.g. https://github.com/pmd/pmd/blob/main/pmd-visualforce/src/test/resources/net/sourceforge/pmd/lang/visualforce/ast/VfExpressionTypeVisitor/metadata/sfdx/objects/Account/fields/Checkbox__c.field-meta.xml - e.g. it definitively has not the root element `CustomField` because it is a standard field.

I'm assuming that the file `.../objects/Case/fields/Description.field-meta.xml` is created and exists when PMD is executed, as otherwise https://github.com/pmd/pmd/blob/d21e5c95d0296b1b53d178f5b344c933eaca92b4/pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/visualforce/ast/ObjectFieldTypes.java#L160 wouldn't return the path to this file.

And I can only reproduce the issue when I extend the sample project we have here at https://github.com/pmd/pmd/tree/main/pmd-visualforce/src/test/resources/net/sourceforge/pmd/lang/visualforce/ast/VfExpressionTypeVisitor/metadata/sfdx by adding a "CaseController.page" and file "objects/Case/fields/Description.field-meta.xml" with some arbitrary xml content. If this file doesn't exist, then no NPE is thrown.

So, that's what I don't understand: Why is this file created, if `Description` is not a custom field?

Could you create a small reproducer project that contains the single apex page which references `Case.Description` with all the folders and files? I really, really want to see a complete project... especially your file "objects/Case/fields/Description.field-meta.xml".

> We created a Salesforce case to inquire whether it’s normal for standard fields to have fewer tags than custom fields (e.g., "type"). They confirmed that this behavior is expected and managed internally.

That's very vague: "...have fewer tags..." - which fields do the standard fields have then? 

We could add a null check to prevent the NPE, but I'd rather not do this without understanding what is happening on the salesforce side there. The code comment would be like "uhh ohh, this is a custom field which is not a custom field, strange".

There are also some standard fields, which appear on all objects, which are handled special. These are documented here: https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/system_fields.htm - well they are called System Fields, but are called Standard Fields here: https://github.com/pmd/pmd/blob/52d88beba4706607603abd06d54e4e1e2100ccb7/pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/visualforce/ast/ObjectFieldTypes.java#L51-L58

But as soon a custom field-meta.xml file is there, we try to read it and will fail for now, if it contains unexpected data.

References when following up on this issue in the future:
* SFDX Project Structure: https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm
  * Decomposed Metadata: https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_decomposed_md_types.htm
* System Fields available on all objects: https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/system_fields.htm
* Standard Objects: https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_list.htm
  * This is represented by https://github.com/apex-dev-tools/sobject-types
* Apex Reference of Built-in classes: https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_ref_guide.htm
  * This is represented by https://github.com/apex-dev-tools/standard-types


I see these steps to go on:
* figure out the definitive content of "objects/Case/fields/Description.field-meta.xml" so, that we know whether we need to lookup a standard field or a custom field
* Add a test case with e.g. a apex page that references Case.Description
* Use apex-dev-tools:sobject-types dependency to lookup standard objects/fields

**Comment by Daniel-Ventura-25** [comment]
Hi @adangel  ,

In regards to your statement:

> When you save your sfdx project as [Salesforce DX Project Structure and Source Format](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm) with decomposed metadata, then there is a file created .../objects/Case/fields/Description.field-meta.xml that has a different content than e.g. https://github.com/pmd/pmd/blob/main/pmd-visualforce/src/test/resources/net/sourceforge/pmd/lang/visualforce/ast/VfExpressionTypeVisitor/metadata/sfdx/objects/Account/fields/Checkbox__c.field-meta.xml - e.g. it definitively has not the root element CustomField because it is a standard field.

If you use this small package example:
```
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>Account</members>
        <name>CustomObject</name>
    </types>
    <version>62.0</version>
</Package>
```

And run the following standard command:
`sf project retrieve start --manifest simple_package_example.xml --target-org [target-org]`

You'll see that even the standard fields for account will start with the root element `CustomField`.

With regards to this: 
> I'm assuming that the file .../objects/Case/fields/Description.field-meta.xml is created and exists when PMD is executed

Yes, you’re right. Since the `.../objects/Case/fields/Description.field-meta.xml` file exists, it triggers the exception.

> So, that's what I don't understand: Why is this file created, if Description is not a custom field?

This file exists because we retrieve all components from the org using a `package.xml` - similar to the example above, but fully complete.

I'm sharing a small project example where we get the error:

[pmdVisualforce.zip](https://github.com/user-attachments/files/19535033/pmdVisualforce.zip)

Thank you!
Daniel Ventura

**Comment by Daniel-Ventura-25** [comment]
Hi @adangel ,

Any news?
